gen/internals.py

```diff
@@ -0,0 +1,464 @@
+
+    def add_variable(self, variable: str):
+        self.variables.add(variable)
+
+    def add_scope(self, scope: Scope):
+        self.sub_scopes[scope.name] = scope
+
+    def __iadd__(self, other):
+        assert isinstance(other, Target), "Internal consistency error, expected Target but got {}".format(type(other))
+        self.variables |= other.variables
+
+        # Add all scopes from the other to this, merging all common sub scopes.
+        for name, scope in other.sub_scopes.items():
+            if name in self.sub_scopes:
+                self.sub_scopes[name] += scope
+            else:
+                self.sub_scopes[name] = scope
+
+        return self
+
+    def __eq__(self, other):
+        assert isinstance(other, Target)
+        return self.variables == other.variables and self.sub_scopes == other.sub_scopes
+
+
+class Source:
+    def __init__(self, entry=None, is_user=False,):
+        self.setters = dict()
+        self.validate = list()
+        self.is_user = is_user
+        if entry:
+            self.add_entry(entry, False)
+
+    def add_setter(self, name, value, is_optional, conditions):
+        self.setters.setdefault(name, list()).append(Setter(name, value, is_optional, conditions, self.is_user))
+
+    def add_conditional_scope(self, scope, conditions):
+        # TODO(cmaloney): 'defaults' are the same as 'can' and 'must' is identical to 'arguments' except
+        # that one takes functions and one takes strings. Simplify to just 'can', 'must'.
+        assert scope.keys() <= {'validate', 'default', 'must', 'conditional'}
+
+        self.validate += scope.get('validate', list())
+
+        for name, fn in scope.get('must', dict()).items():
+            self.add_setter(name, fn, False, conditions)
+
+        for name, fn in scope.get('default', dict()).items():
+            self.add_setter(name, fn, True, conditions)
+
+        for name, cond_options in scope.get('conditional', dict()).items():
+            for value, sub_scope in cond_options.items():
+                self.add_conditional_scope(sub_scope, conditions + [(name, value)])
+
+    add_must = partialmethod(add_setter, is_optional=False, conditions=[])
+
+    def add_value_dict(self, value_dict):
+        for name, value in value_dict.items():
+            self.add_must(name, value)
+
+    def remove_setters(self, scope):
+        def del_setter(name):
+            if name in self.setters:
+                del self.setters[name]
+
+        for name in scope.get('must', dict()).keys():
+            del_setter(name)
+
+        for name in scope.get('default', dict()).keys():
+            del_setter(name)
+
+        for name, cond_options in scope.get('conditional', dict()).items():
+            if name in self.setters:
+                raise NotImplementedError("Should conditional setters overwrite all setters?")
+
+    def add_entry(self, entry, replace_existing):
+        if replace_existing:
+            self.remove_setters(entry)
+
```

Comment:
I'm a fan of leaving vertical whitespace between things (esp. after ifs). Practically things like the whitespace comes down to a whole lot of personal preference, so I tend to try and avoid it in code reviews (our codebase isn't consistent, and trying to get it consistent isn't of great value). Tox / code quality checks enforce the major whitespace bits that are cared about / required for landing.